File: app/src/main/java/com/runtracker/app/ui/screens/wellness/WellnessViewModel.kt
package com.runtracker.app.ui.screens.wellness

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.runtracker.shared.data.model.*
import com.runtracker.shared.data.repository.GamificationRepository
import com.runtracker.shared.data.repository.MentalHealthRepository
import com.runtracker.shared.data.repository.RunRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import com.runtracker.shared.util.TimeUtils
import java.util.*
import javax.inject.Inject

@HiltViewModel
class WellnessViewModel @Inject constructor(
    private val runRepository: RunRepository,
    private val gamificationRepository: GamificationRepository,
    private val mentalHealthRepository: MentalHealthRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow(WellnessUiState())
    val uiState: StateFlow<WellnessUiState> = _uiState.asStateFlow()

    init {
        loadWellnessData()
    }

    private fun loadWellnessData() {
        viewModelScope.launch {
            // Load gamification data
            gamificationRepository.getUserGamification().collect { gamification ->
                _uiState.update { it.copy(gamification = gamification) }
                calculateOverallWellness()
            }
        }
        
        viewModelScope.launch {
            // Load today's rings
            val rings = gamificationRepository.getTodayRings()
            _uiState.update { it.copy(todayRings = rings) }
            calculateOverallWellness()
        }
        
        viewModelScope.launch {
            // Load recent wellness check-ins
            mentalHealthRepository.getRecentWellnessCheckins(7).collect { checkins ->
                _uiState.update { it.copy(recentCheckins = checkins) }
                calculateOverallWellness()
            }
        }
        
        viewModelScope.launch {
            // Load today's check-in
            val todayCheckin = mentalHealthRepository.getTodayCheckin()
            _uiState.update { it.copy(todayCheckin = todayCheckin) }
            calculateOverallWellness()
        }
        
        viewModelScope.launch {
            // Load recent runs for training load
            val weekAgo = System.currentTimeMillis() - TimeUtils.ONE_WEEK_MS
            runRepository.getRunsSince(weekAgo).collect { runs ->
                _uiState.update { it.copy(recentRuns = runs) }
                calculateTrainingLoad(runs)
                calculateOverallWellness()
            }
        }
        
        viewModelScope.launch {
            // Load mindfulness stats
            val mindfulnessMinutes = mentalHealthRepository.getTotalMindfulnessMinutesThisWeek()
            _uiState.update { it.copy(mindfulnessMinutesThisWeek = mindfulnessMinutes) }
        }
    }

    private fun calculateTrainingLoad(runs: List<Run>) {
        if (runs.isEmpty()) {
            _uiState.update { it.copy(trainingLoad = TrainingLoad()) }
            return
        }
        
        val now = System.currentTimeMillis()
        val weekAgo = now - TimeUtils.ONE_WEEK_MS
        val twoWeeksAgo = now - (2 * TimeUtils.ONE_WEEK_MS)
        
        // Acute load (last 7 days)
        val acuteRuns = runs.filter { it.startTime >= weekAgo }
        val acuteDistance = acuteRuns.sumOf { it.distanceMeters }
        val acuteDuration = acuteRuns.sumOf { it.durationMillis }
        
        // Calculate training stress score (simplified)
        val acuteLoad = acuteRuns.sumOf { run ->
            val durationHours = run.durationMillis / 3600000.0
            val intensityFactor = when {
                run.avgPaceSecondsPerKm > 0 && run.avgPaceSecondsPerKm < 300 -> 1.5 // Fast pace
                run.avgPaceSecondsPerKm > 0 && run.avgPaceSecondsPerKm < 360 -> 1.2 // Moderate pace
                else -> 1.0 // Easy pace
            }
            (durationHours * intensityFactor * 100).toInt()
        }
        
        // Chronic load (average of last 4 weeks, simplified to 2 weeks here)
        val chronicLoad = (acuteLoad * 0.7).toInt() // Simplified calculation
        
        // Acute:Chronic ratio
        val acwr = if (chronicLoad > 0) acuteLoad.toFloat() / chronicLoad else 1f
        
        // Training status based on ACWR
        val status = when {
            acwr < 0.8 -> TrainingStatus.DETRAINING
            acwr in 0.8..1.3 -> TrainingStatus.OPTIMAL
            acwr in 1.3..1.5 -> TrainingStatus.HIGH_RISK
            else -> TrainingStatus.VERY_HIGH_RISK
        }
        
        _uiState.update { 
            it.copy(
                trainingLoad = TrainingLoad(
                    acuteLoad = acuteLoad,
                    chronicLoad = chronicLoad,
                    acuteChronicRatio = acwr,
                    status = status,
                    weeklyDistance = acuteDistance,
                    weeklyDuration = acuteDuration,
                    weeklyRuns = acuteRuns.size
                )
            ) 
        }
    }

    private fun calculateOverallWellness() {
        val state = _uiState.value
        
        var score = 50 // Base score
        var factors = 0
        
        // Training consistency (from gamification)
        state.gamification?.let { g ->
            factors++
            score += when {
                g.currentStreak >= 7 -> 15
                g.currentStreak >= 3 -> 10
                g.currentStreak >= 1 -> 5
                else -> -5
            }
        }
        
        // Daily activity (from rings)
        state.todayRings?.let { rings ->
            factors++
            val ringCompletion = (rings.moveProgress + rings.exerciseProgress) / 2
            score += (ringCompletion * 15).toInt()
        }
        
        // Readiness from check-in
        state.todayCheckin?.readinessScore?.let { readiness ->
            factors++
            score += ((readiness - 50) / 5) // -10 to +10
        }
        
        // Training load balance
        state.trainingLoad.let { load ->
            if (load.acuteLoad > 0) {
                factors++
                score += when (load.status) {
                    TrainingStatus.OPTIMAL -> 10
                    TrainingStatus.DETRAINING -> -5
                    TrainingStatus.HIGH_RISK -> -5
                    TrainingStatus.VERY_HIGH_RISK -> -15
                }
            }
        }
        
        // Mindfulness bonus
        if (state.mindfulnessMinutesThisWeek > 0) {
            factors++
            score += minOf(state.mindfulnessMinutesThisWeek / 2, 10)
        }
        
        val finalScore = score.coerceIn(0, 100)
        val status = when {
            finalScore >= 80 -> WellnessStatus.EXCELLENT
            finalScore >= 60 -> WellnessStatus.GOOD
            finalScore >= 40 -> WellnessStatus.FAIR
            else -> WellnessStatus.NEEDS_ATTENTION
        }
        
        _uiState.update { 
            it.copy(
                overallWellnessScore = finalScore,
                wellnessStatus = status,
                recommendations = generateRecommendations(state, finalScore)
            ) 
        }
    }

    private fun generateRecommendations(state: WellnessUiState, score: Int): List<WellnessRecommendation> {
        val recommendations = mutableListOf<WellnessRecommendation>()
        
        // Check-in recommendation
        if (state.todayCheckin == null) {
            recommendations.add(
                WellnessRecommendation(
                    icon = "📋",
                    title = "Complete Daily Check-in",
                    description = "Log your sleep, mood, and energy to get personalized insights",
                    priority = RecommendationPriority.HIGH
                )
            )
        }
        
        // Streak recommendation
        state.gamification?.let { g ->
            if (g.currentStreak == 0) {
                recommendations.add(
                    WellnessRecommendation(
                        icon = "🔥",
                        title = "Start Your Streak",
                        description = "Complete a workout today to begin building consistency",
                        priority = RecommendationPriority.MEDIUM
                    )
                )
            }
        }
        
        // Training load recommendations
        when (state.trainingLoad.status) {
            TrainingStatus.DETRAINING -> {
                recommendations.add(
                    WellnessRecommendation(
                        icon = "📈",
                        title = "Increase Training",
                        description = "Your training load is low. Consider adding more workouts this week.",
                        priority = RecommendationPriority.MEDIUM
                    )
                )
            }
            TrainingStatus.HIGH_RISK, TrainingStatus.VERY_HIGH_RISK -> {
                recommendations.add(
                    WellnessRecommendation(
                        icon = "⚠️",
                        title = "Recovery Needed",
                        description = "Your training load is high. Consider a rest day or easy workout.",
                        priority = RecommendationPriority.HIGH
                    )
                )
            }
            else -> {}
        }
        
        // Mindfulness recommendation
        if (state.mindfulnessMinutesThisWeek < 10) {
            recommendations.add(
                WellnessRecommendation(
                    icon = "🧘",
                    title = "Add Mindfulness",
                    description = "Try a breathing exercise to reduce stress and improve recovery",
                    priority = RecommendationPriority.LOW
                )
            )
        }
        
        // Sleep recommendation
        state.todayCheckin?.let { checkin ->
            checkin.sleepHours?.let { hours ->
                if (hours < 7) {
                    recommendations.add(
                        WellnessRecommendation(
                            icon = "😴",
                            title = "Prioritize Sleep",
                            description = "You got ${hours}h of sleep. Aim for 7-9 hours for optimal recovery.",
                            priority = RecommendationPriority.HIGH
                        )
                    )
                }
            }
        }
        
        return recommendations.sortedBy { it.priority.ordinal }
    }

    fun refresh() {
        loadWellnessData()
    }
}

data class WellnessUiState(
    val overallWellnessScore: Int = 50,
    val wellnessStatus: WellnessStatus = WellnessStatus.FAIR,
    val gamification: UserGamification? = null,
    val todayRings: DailyRings? = null,
    val todayCheckin: WellnessCheckin? = null,
    val recentCheckins: List<WellnessCheckin> = emptyList(),
    val recentRuns: List<Run> = emptyList(),
    val trainingLoad: TrainingLoad = TrainingLoad(),
    val mindfulnessMinutesThisWeek: Int = 0,
    val recommendations: List<WellnessRecommendation> = emptyList(),
    val isLoading: Boolean = false
)

data class TrainingLoad(
    val acuteLoad: Int = 0,
    val chronicLoad: Int = 0,
    val acuteChronicRatio: Float = 1f,
    val status: TrainingStatus = TrainingStatus.OPTIMAL,
    val weeklyDistance: Double = 0.0,
    val weeklyDuration: Long = 0,
    val weeklyRuns: Int = 0
)

enum class TrainingStatus(val label: String, val color: Long) {
    DETRAINING("Detraining", 0xFFFF9800),
    OPTIMAL("Optimal", 0xFF4CAF50),
    HIGH_RISK("High Risk", 0xFFFF9800),
    VERY_HIGH_RISK("Very High Risk", 0xFFF44336)
}

enum class WellnessStatus(val label: String, val emoji: String) {
    EXCELLENT("Excellent", "🌟"),
    GOOD("Good", "✅"),
    FAIR("Fair", "😐"),
    NEEDS_ATTENTION("Needs Attention", "⚠️")
}

data class WellnessRecommendation(
    val icon: String,
    val title: String,
    val description: String,
    val priority: RecommendationPriority
)

enum class RecommendationPriority {
    HIGH, MEDIUM, LOW
}

File: iosApp/GoSteady/Sources/ViewModels/WellnessViewModel.swift
import Foundation
import Observation

// MARK: - Wellness Enums

enum WellnessStatus: String, CaseIterable {
    case excellent, good, fair, needsAttention

    var label: String {
        switch self {
        case .excellent: return "Excellent"
        case .good: return "Good"
        case .fair: return "Fair"
        case .needsAttention: return "Needs Attention"
        }
    }

    var emoji: String {
        switch self {
        case .excellent: return "star.fill"
        case .good: return "checkmark.circle.fill"
        case .fair: return "minus.circle.fill"
        case .needsAttention: return "exclamationmark.triangle.fill"
        }
    }

    var colorHex: String {
        switch self {
        case .excellent: return "4CAF50"
        case .good: return "8BC34A"
        case .fair: return "FF9800"
        case .needsAttention: return "F44336"
        }
    }
}

enum WellnessTrainingStatus: String, CaseIterable {
    case detraining, optimal, highRisk, veryHighRisk

    var label: String {
        switch self {
        case .detraining: return "Detraining"
        case .optimal: return "Optimal"
        case .highRisk: return "High Risk"
        case .veryHighRisk: return "Very High Risk"
        }
    }

    var colorHex: String {
        switch self {
        case .detraining: return "FF9800"
        case .optimal: return "4CAF50"
        case .highRisk: return "FF9800"
        case .veryHighRisk: return "F44336"
        }
    }
}

enum RecommendationPriority: Int, Comparable {
    case high = 0, medium = 1, low = 2

    static func < (lhs: RecommendationPriority, rhs: RecommendationPriority) -> Bool {
        lhs.rawValue < rhs.rawValue
    }
}

struct WellnessTrainingLoad {
    var acuteLoad: Int = 0
    var chronicLoad: Int = 0
    var acuteChronicRatio: Float = 1.0
    var status: WellnessTrainingStatus = .optimal
    var weeklyDistanceMeters: Double = 0
    var weeklyDurationMinutes: Int = 0
    var weeklyRuns: Int = 0
}

struct WellnessRecommendation: Identifiable {
    let id = UUID()
    var icon: String
    var title: String
    var description: String
    var priority: RecommendationPriority
}

struct WellnessSleepData: Identifiable {
    let id = UUID()
    var date: Date
    var hours: Double
    var quality: Int // 1-5
}

struct HRVDataPoint: Identifiable {
    let id = UUID()
    var date: Date
    var value: Double // ms
}

struct RestingHRDataPoint: Identifiable {
    let id = UUID()
    var date: Date
    var bpm: Int
}

struct ReadinessBreakdown {
    var sleepScore: Int = 0
    var recoveryScore: Int = 0
    var stressScore: Int = 0
    var trainingLoadScore: Int = 0
}

// MARK: - WellnessViewModel

@Observable
final class WellnessViewModel {
    // Dependencies
    private let wellnessRepository: WellnessRepository?
    private let healthKitService: HealthKitService
    private let smartTrainerService: SmartTrainerService?
    private let gamificationRepository: GamificationRepository?

    // Overall State
    var overallWellnessScore: Int = 50
    var wellnessStatus: WellnessStatus = .fair
    var isLoading = true
    var errorMessage: String?

    // Readiness
    var readinessScore: Int = 0
    var readinessBreakdown = ReadinessBreakdown()

    // Sleep
    var lastNightSleepHours: Double?
    var sleepQuality: Int?
    var weeklyWellnessSleepData: [WellnessSleepData] = []

    // HRV
    var currentHRV: Double?
    var weeklyHRVData: [HRVDataPoint] = []
    var hrvBaseline: Double = 45.0

    // Resting HR
    var restingHR: Int?
    var weeklyRestingHR: [RestingHRDataPoint] = []

    // Stress
    var stressLevel: Int = 3 // 1-5

    // Training Load
    var trainingLoad = WellnessTrainingLoad()

    // Streaks / Gamification
    var currentStreak: Int = 0
    var mindfulnessMinutesThisWeek: Int = 0

    // Recommendations
    var recommendations: [WellnessRecommendation] = []

    // Weekly trend data
    var weeklyReadinessScores: [(date: Date, score: Int)] = []

    init(
        wellnessRepository: WellnessRepository? = nil,
        healthKitService: HealthKitService = HealthKitService(),
        smartTrainerService: SmartTrainerService? = nil,
        gamificationRepository: GamificationRepository? = nil
    ) {
        self.wellnessRepository = wellnessRepository
        self.healthKitService = healthKitService
        self.smartTrainerService = smartTrainerService
        self.gamificationRepository = gamificationRepository
        loadData()
    }

    func loadData() {
        isLoading = true
        Task { @MainActor in
            // Load sleep data
            let sleep = await healthKitService.fetchLastNightSleep()
            self.lastNightSleepHours = sleep?.totalHours
            self.sleepQuality = sleep.map { sleepData in
                switch sleepData.totalHours {
                case 8...: return 5
                case 7...: return 4
                case 6...: return 3
                case 5...: return 2
                default: return 1
                }
            }

            // Load HRV
            let hrv = await healthKitService.fetchLatestHRV()
            self.currentHRV = hrv

            // Load resting HR
            let rhr = await healthKitService.fetchRestingHeartRate()
            self.restingHR = rhr

            // Load readiness
            if let smartTrainer = smartTrainerService {
                let readiness = await smartTrainer.calculateReadiness()
                self.readinessScore = Int(readiness.overall)
                self.readinessBreakdown = ReadinessBreakdown(
                    sleepScore: Int(readiness.sleepScore),
                    recoveryScore: Int(readiness.hrvScore),
                    stressScore: Int(readiness.moodScore),
                    trainingLoadScore: Int(readiness.recentLoadScore)
                )
            }

            // Load streak
            if let gamification = gamificationRepository {
                let userGamification = gamification.getOrCreateUserGamification()
                self.currentStreak = userGamification.currentStreak
            }

            // Load mindfulness minutes
            if let repo = wellnessRepository {
                self.mindfulnessMinutesThisWeek = repo.totalMindfulnessMinutesThisWeek()

                let checkins = repo.fetchRecentCheckins(limit: 7)
                self.weeklyReadinessScores = checkins.compactMap { checkin in
                    if let score = checkin.readinessScore {
                        return (checkin.date, score)
                    }
                    return nil
                }
            }

            calculateOverallWellness()
            generateRecommendations()
            self.isLoading = false
        }
    }

    func refresh() {
        loadData()
    }

    private func calculateOverallWellness() {
        var score = 50
        var factors = 0

        // Streak contribution
        if currentStreak >= 7 {
            score += 15; factors += 1
        } else if currentStreak >= 3 {
            score += 10; factors += 1
        } else if currentStreak >= 1 {
            score += 5; factors += 1
        } else {
            score -= 5; factors += 1
        }

        // Readiness contribution
        if readinessScore > 0 {
            score += (readinessScore - 50) / 5
            factors += 1
        }

        // Training load balance
        if trainingLoad.acuteLoad > 0 {
            switch trainingLoad.status {
            case .optimal: score += 10
            case .detraining: score -= 5
            case .highRisk: score -= 5
            case .veryHighRisk: score -= 15
            }
            factors += 1
        }

        // Sleep contribution
        if let hours = lastNightSleepHours {
            if hours >= 7 {
                score += 10
            } else if hours >= 6 {
                score += 5
            } else {
                score -= 10
            }
            factors += 1
        }

        // Mindfulness bonus
        if mindfulnessMinutesThisWeek > 0 {
            score += min(mindfulnessMinutesThisWeek / 2, 10)
            factors += 1
        }

        let finalScore = max(0, min(100, score))
        overallWellnessScore = finalScore

        if finalScore >= 80 {
            wellnessStatus = .excellent
        } else if finalScore >= 60 {
            wellnessStatus = .good
        } else if finalScore >= 40 {
            wellnessStatus = .fair
        } else {
            wellnessStatus = .needsAttention
        }
    }

    private func generateRecommendations() {
        var recs: [WellnessRecommendation] = []

        // Sleep
        if let hours = lastNightSleepHours, hours < 7 {
            recs.append(WellnessRecommendation(
                icon: "moon.zzz.fill",
                title: "Prioritize Sleep",
                description: "You got \(String(format: "%.1f", hours))h of sleep. Aim for 7-9 hours for optimal recovery.",
                priority: .high
            ))
        }

        // Training load
        switch trainingLoad.status {
        case .detraining:
            recs.append(WellnessRecommendation(
                icon: "chart.line.uptrend.xyaxis",
                title: "Increase Training",
                description: "Your training load is low. Consider adding more workouts this week.",
                priority: .medium
            ))
        case .highRisk, .veryHighRisk:
            recs.append(WellnessRecommendation(
                icon: "exclamationmark.triangle.fill",
                title: "Recovery Needed",
                description: "Your training load is high. Consider a rest day or easy workout.",
                priority: .high
            ))
        case .optimal:
            break
        }

        // Mindfulness
        if mindfulnessMinutesThisWeek < 10 {
            recs.append(WellnessRecommendation(
                icon: "brain.head.profile",
                title: "Add Mindfulness",
                description: "Try a breathing exercise to reduce stress and improve recovery.",
                priority: .low
            ))
        }

        // HRV trend
        if let hrv = currentHRV, hrv < hrvBaseline * 0.8 {
            recs.append(WellnessRecommendation(
                icon: "heart.text.square",
                title: "HRV Below Baseline",
                description: "Your HRV is lower than usual. Consider lighter training today.",
                priority: .high
            ))
        }

        // Streak
        if currentStreak == 0 {
            recs.append(WellnessRecommendation(
                icon: "flame",
                title: "Start Your Streak",
                description: "Complete a workout today to begin building consistency.",
                priority: .medium
            ))
        }

        recommendations = recs.sorted { $0.priority < $1.priority }
    }

    // Computed properties for detail screen
    var hrvTrendDirection: String {
        guard weeklyHRVData.count >= 2 else { return "Stable" }
        let recent = weeklyHRVData.suffix(3).map(\.value)
        let older = weeklyHRVData.prefix(3).map(\.value)
        let recentAvg = recent.reduce(0, +) / Double(recent.count)
        let olderAvg = older.reduce(0, +) / Double(older.count)
        if recentAvg > olderAvg * 1.05 { return "Improving" }
        if recentAvg < olderAvg * 0.95 { return "Declining" }
        return "Stable"
    }

    var sleepAverage: Double {
        guard !weeklyWellnessSleepData.isEmpty else { return 0 }
        return weeklyWellnessSleepData.map(\.hours).reduce(0, +) / Double(weeklyWellnessSleepData.count)
    }

    var acuteChronicRatioFormatted: String {
        String(format: "%.2f", trainingLoad.acuteChronicRatio)
    }
}

